### ISSUE TITLE
Malformed classifications being saved that break exports

### ISSUE THREAD

**Issue by zwolf** [zooniverse/Panoptes-Front-End#4401]
Occasionally, classification export jobs for a given project or workflow will mysteriously vanish and unhelpful NoMethodErrors show up in Honeybadger. This is due to malformed classifications being saved by PFE and when the exporter tries to untangle them, it chokes and dies silently. This seems to happen with huge tasks consisting of many dropdowns and combos. 

I've tracked down a few examples which will hopefully help us track this down. I've formatted them but put them in gists because lordy they're humungous. Project 3414. workflow 5919. There's a significant number of these classifications for this project/workflow and tbh I'm not sure yet how we fix/clean these so that exports run once we stop this from happening.

Workflow task definition:
https://gist.github.com/zwolf/f00b4f67629dc465aecbe8be9ffd9fc8

Here's an annotation that causes the NoMethodError in the exporter:
https://gist.github.com/zwolf/ed952817b4308ca3d64c6c972d299093

Seems like this kind of thing could be the culprit:
```
      {
        "task": "T13",
        "value": [
          null,
          null,
          {
            "value": "Canada",
            "option": false
          }
        ]
      },
```

cc: @mcbouslog / @trouille

**Comment by eatyourgreens** [comment]
If I’m reading that correctly, Country has been selected but not State or County? Does the dropdown menu default to null if nothing is selected? I know that’s used as the default value for other tasks.

**Comment by zwolf** [comment]
I don't think null values are problematic unless they themselves are dropdowns? Hard to tell because the workflow is so complicated. Here's the code that does the pulling apart: https://github.com/zooniverse/Panoptes/blob/master/lib/formatter/csv/annotation_for_csv.rb#L101 and the actual error occurs on line 124.

Here's another even less clear example of a busted annotation:

```[
  {
    "task": "T6",
    "value": 0
  },
  {
    "task": "T1",
    "value": [
      {
        "task": "T5",
        "value": "km "
      },
      {
        "task": "T7",
        "value": [
          {
            "value": null,
            "option": false
          }
        ]
      },
      {
        "task": "T8",
        "value": ""
      },
      {
        "task": "T9",
        "value": [
          {
            "value": null,
            "option": false
          },
          {
            "value": null,
            "option": false
          }
        ]
      }
    ]
  },
  {
    "task": "T2",
    "value": [
      {
        "task": "T10",
        "value": ""
      },
      {
        "task": "T11",
        "value": ""
      },
      {
        "task": "T12",
        "value": ""
      },
      {
        "task": "T13",
        "value": []
      },
      {
        "task": "T14",
        "value": ""
      }
    ]
  }
]```

**Comment by mcbouslog** [comment]
The problem in the first example is that the dropdown task component isn't setting default annotations if the component is updated; it's only checking/setting default annotations [on component mount](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/dropdown/index.cjsx#L95).

So it appears that with a multi-view workflow and a dropdown task in the same spot in view 1 and 2, the dropdown component would not set default annotations, leaving `null`s instead of desired `{ option: false, value: null }`'s.

Compare the initial dropdown annotation for task T0 (Dropdown2-3, second view) in these workflows:
- currently: https://master.pfe-preview.zooniverse.org/projects/markb-panoptes/test-project-1-mb/classify?reload=0&workflow=3165
- fix: https://dropdown-default-update-fix.pfe-preview.zooniverse.org/projects/markb-panoptes/test-project-1-mb/classify?reload=1&workflow=3165

I'll look closer at the second example tomorrow.

**Comment by mcbouslog** [comment]
@zwolf - is the second example from the same project/workflow (3414/5919)?

**Comment by zwolf** [comment]
Yes, same workflow.

**Comment by mcbouslog** [comment]
I think the empty array in task T13 was indeed what caused the issue in the second example too Zach, just as you predicted in our conversation.

The fix in PR #4404 should address that scenario (empty array causing error in data processing) as well as the first scenario (array with null items).
